File: frontend/src/api/trading.ts
import api from './index'

// --- Types ---

export interface Account {
  id: string
  user_id: string
  name: string
  market: string
  broker: string | null
  account_no: string | null
  balance: number
  is_simulated: boolean
  created_at: string
}

export interface Position {
  id: string
  account_id: string
  symbol: string
  name: string | null
  quantity: number
  available_quantity: number
  avg_cost: number
  current_price: number
  market_value: number
  unrealized_pnl: number
  realized_pnl: number
  updated_at: string
}

export interface OrderRecord {
  id: string
  account_id: string
  symbol: string
  name: string | null
  side: string
  order_type: string
  status: string
  quantity: number
  filled_quantity: number
  price: number | null
  filled_price: number | null
  commission: number
  submitted_at: string | null
  filled_at: string | null
  cancelled_at: string | null
  remark: string | null
  created_at: string
}

export interface OrderCreateRequest {
  account_id: string
  symbol: string
  side: 'buy' | 'sell'
  order_type: 'market' | 'limit'
  quantity: number
  price?: number
  remark?: string
}

export interface AccountCreateRequest {
  name: string
  market: string
  is_simulated?: boolean
  balance?: number
}

// --- Account APIs ---

export async function createAccount(data: AccountCreateRequest): Promise<Account> {
  const resp = await api.post<Account>('/v1/accounts', data)
  return resp.data
}

export async function listAccounts(): Promise<Account[]> {
  const resp = await api.get<Account[]>('/v1/accounts')
  return resp.data
}

export async function getAccount(id: string): Promise<Account> {
  const resp = await api.get<Account>(`/v1/accounts/${id}`)
  return resp.data
}

export async function listPositions(accountId: string): Promise<Position[]> {
  const resp = await api.get<Position[]>(`/v1/accounts/${accountId}/positions`)
  return resp.data
}

export async function resetAccount(id: string): Promise<Account> {
  const resp = await api.post<Account>(`/v1/accounts/${id}/reset`)
  return resp.data
}

// --- Order APIs ---

export async function placeOrder(data: OrderCreateRequest): Promise<OrderRecord> {
  const resp = await api.post<OrderRecord>('/v1/trading/orders', data)
  return resp.data
}

export interface OrderListResponse {
  items: OrderRecord[]
  total: number
}

export async function listOrders(
  accountId: string,
  status?: string,
  symbol?: string,
  limit = 20,
  offset = 0,
): Promise<OrderListResponse> {
  const resp = await api.get<OrderListResponse>('/v1/trading/orders', {
    params: { account_id: accountId, status, symbol, limit, offset },
  })
  return resp.data
}

export async function getOrder(id: string): Promise<OrderRecord> {
  const resp = await api.get<OrderRecord>(`/v1/trading/orders/${id}`)
  return resp.data
}

export async function cancelOrder(id: string): Promise<OrderRecord> {
  const resp = await api.delete<OrderRecord>(`/v1/trading/orders/${id}`)
  return resp.data
}

File: backend/app/schemas/trading.py
"""Trading request/response schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Account ---


class AccountCreate(BaseModel):
    name: str = Field(max_length=100)
    market: str  # a_share, us_stock, hk_stock, fund, bond
    is_simulated: bool = True
    balance: Decimal | None = None  # None → use default sim balance


class AccountRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    market: str
    broker: str | None
    account_no: str | None
    balance: Decimal
    is_simulated: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class AccountWithPositions(AccountRead):
    positions: list[PositionRead] = []


# --- Position ---


class PositionRead(BaseModel):
    id: uuid.UUID
    account_id: uuid.UUID
    symbol: str
    name: str | None
    quantity: int
    available_quantity: int
    avg_cost: Decimal
    current_price: Decimal
    market_value: Decimal
    unrealized_pnl: Decimal
    realized_pnl: Decimal
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Order ---


class OrderCreate(BaseModel):
    account_id: uuid.UUID
    symbol: str = Field(max_length=20)
    side: str  # buy, sell
    order_type: str  # market, limit
    quantity: int = Field(gt=0)
    price: Decimal | None = None  # required for limit orders
    remark: str | None = Field(None, max_length=500)


class OrderRead(BaseModel):
    id: uuid.UUID
    account_id: uuid.UUID
    symbol: str
    name: str | None
    side: str
    order_type: str
    status: str
    quantity: int
    filled_quantity: int
    price: Decimal | None
    filled_price: Decimal | None
    commission: Decimal
    submitted_at: datetime | None
    filled_at: datetime | None
    cancelled_at: datetime | None
    broker_order_id: str | None
    remark: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Performance ---


class PerformanceStats(BaseModel):
    total_value: Decimal
    total_cost: Decimal
    total_pnl: Decimal
    total_pnl_pct: Decimal
    realized_pnl: Decimal
    unrealized_pnl: Decimal
    win_count: int
    loss_count: int
    win_rate: Decimal | None


# Forward ref update
AccountWithPositions.model_rebuild()
